---
title: "lab7_cappy"
format: html
editor: visual
---

# GWAS tutorial

Once you are done setting up your directory, your task for the day is to get GWAS running in `R` using its example data. The guide is found [here](https://cran.r-project.org/web/packages/statgenGWAS/vignettes/GWAS.html). Follow along until you get to the Manhattan Plots, but use the "plant.height" trait rather than "grain.yield", and send me your result.

```{r}
library("statgenGWAS")
# load in the data
data(dropsMarkers)
data(dropsMap)
data(dropsPheno)

## Add genotypes as row names of dropsMarkers and drop Ind column.
rownames(dropsMarkers) <- dropsMarkers[["Ind"]]
dropsMarkers <- dropsMarkers[colnames(dropsMarkers) != "Ind"]

## Add genotypes as row names of dropsMap.
rownames(dropsMap) <- dropsMap[["SNP.names"]]
## Rename Chomosome and Position columns.
colnames(dropsMap)[match(c("Chromosome", "Position"), colnames(dropsMap))] <- c("chr", "pos")

## Create a gData object containing map and marker information.
gDataDrops <- createGData(geno = dropsMarkers, map = dropsMap)

## Rename Variety_ID to genotype.
colnames(dropsPheno)[colnames(dropsPheno) == "Variety_ID"] <- "genotype"
## Select relevant columns and convert data to a list.
dropsPhenoList <- split(x = dropsPheno[c("genotype", "grain.yield",
                                         "grain.number", "seed.size",
                                         "anthesis", "silking", "plant.height",
                                         "tassel.height", "ear.height")], 
                        f = dropsPheno[["Experiment"]])
## Add phenotypic data to gDataDrops.
gDataDrops <- createGData(gData = gDataDrops, pheno = dropsPhenoList)

## Summarize gDataDrops.
summary(gDataDrops, trials = "Mur13W")

## Plot genetic map.
plot(gDataDrops)

## Plot genetic map.
## Highlight the 20.000th marker in the map.
plot(gDataDrops, highlight = dropsMap[20000, ])

## Remove duplicate SNPs from gDataDrops.
gDataDropsDedup <- codeMarkers(gDataDrops, impute = FALSE, verbose = TRUE) 

## Copy gData object.
gDataDropsMiss <- gDataDrops
## Add random missing values to 1% of the values in the marker matrix.
set.seed(1)
nVal <- nrow(gDataDropsMiss$markers) * ncol(gDataDropsMiss$markers)
gDataDropsMiss$markers[sample(x = 1:nVal, size = nVal / 100)] <- NA


## Run single trait GWAS for traits 'plant.height' and 'anthesis' for trial Mur13W.
GWASDrops <- runSingleTraitGwas(gData = gDataDropsDedup,
                                trials = "Mur13W",
                                traits = c("plant.height", "anthesis"))

# print(head(GWASDrops$GWAResult$Mur13W), row.names = FALSE)
# print(GWASDrops$signSnp$Mur13W, row.names = FALSE)
    # get null, but that's ok --> working with different traits than actual tutorial
# summary(GWASDrops)
```

## Plots

```{r}
## Plot a QQ-plot of GWAS Drops.
plot(GWASDrops, plotType = "qq", trait = "plant.height")
```

![](images/paste-1.png)

```{r}
## Plot a manhattan plot of GWAS Drops.
plot(GWASDrops, plotType = "manhattan", trait = "plant.height")
```

![](images/paste-2.png)